src/codec.rs

```diff
@@ -0,0 +1,303 @@
+//! Implementation of the DAG-JOSE code by defining DAG-CBOR structure.
+#![deny(missing_docs)]
+#![deny(warnings)]
+
+use libipld::{Cid, Ipld};
+use serde_derive::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+use crate::{bytes::Bytes, error::Error, JsonWebEncryption};
+use crate::{Jose, Signature};
+use crate::{JsonWebSignature, Recipient};
+
+/// Encoded represents the union of fields from JSON Web Signature (JWS)
+/// and JSON Web Encryption (JWE) objects.
+///
+/// The data is repsented as bytes and therefore can be encoded
+/// into a DAG-JOSE object using DAG-CBOR.
+///
+/// See https://ipld.io/specs/codecs/dag-jose/spec/#format
+#[derive(PartialEq, Default, Debug, Serialize, Deserialize)]
+pub struct Encoded {
+    // NOTE: Serde serialized fields in the order they appear
+    // However DAG-CBOR specifies the fields must be serialized
+    // sorted by length of key.
+    //
+    // Within each grouping the fields are defined in their correct
+    // sort order.
+
+    // JWS fields
+    #[serde(skip_serializing_if = "Option::is_none")]
+    payload: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    signatures: Option<Vec<EncodedSignature>>,
+
+    // JWE fields
+    #[serde(skip_serializing_if = "Option::is_none")]
+    iv: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    aad: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    tag: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    protected: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    ciphertext: Option<Bytes>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    recipients: Option<Vec<EncodedRecipient>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    unprotected: Option<BTreeMap<String, Ipld>>,
+}
+
+impl<'a> TryFrom<&'a JsonWebSignature> for Encoded {
+    type Error = Error;
+
+    fn try_from(value: &'a JsonWebSignature) -> Result<Self, Self::Error> {
+        Ok(Self {
+            payload: Some(Bytes::from_base64(value.payload.as_str())?),
+            signatures: if value.signatures.is_empty() {
+                None
+            } else {
+                Some(
+                    value
+                        .signatures
+                        .iter()
+                        .map(EncodedSignature::try_from)
+                        .collect::<Result<Vec<EncodedSignature>, Self::Error>>()?,
+                )
+            },
+            aad: None,
+            ciphertext: None,
+            iv: None,
+            protected: None,
+            recipients: None,
+            tag: None,
+            unprotected: None,
+        })
+    }
+}
+
+impl TryFrom<Encoded> for JsonWebSignature {
```

Comment:
Similar comment as before, do we want to reduce implementations by using `try_into()`? Do we have a number of cases where it makes sense to use the reference?